File: Napakalakirb/lib/numeric_bad_consequence.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

class NumericBadConsequence < BadConsequence
  
  attr_reader :nVisibleTreasures
  attr_reader :nHiddenTreasures
  
  def initialize(aText, someLevels,someVisibleTreasures, someHiddenTreasures)
    super(aText,someLevels)
    @nVisibleTreasures = someVisibleTreasures
    @nHiddenTreasures = someHiddenTreasures
  end
  
  def isEmpty
    if( @nVisibleTreasures == 0 && @nHiddenTreasures == 0 )
      return true
    end
    return false
  end
  
  def substractVisibleTreasure( t ) 
    if ( @nVisibleTreasures > 0 )
      @nVisibleTreasures -= 1
    end
  end
  
  def substractHiddenTreasure( t )
    if ( @nHiddenTreasures > 0 )
        @nHiddenTreasures -= 1
    end
  end
  
  def adjustToFitTreasureLists( v , h )
    if( @nVisibleTreasures == 0 )
      nv = 0;
    else
      nv = v.size
    end
    if ( @nHiddenTreasures == 0 )
      nh = 0
    else
      nh = h.size
    end
    res = NumericBadConsequence.new(@text,0,nv,nh);
    return res;
  end
  
  def to_s
    return super() + " NºTesoros visibles: #{@nVisibleTreasures},
     NºTesoros ocultos: #{@nHiddenTreasures}"
  end
  
end

File: Napakalaki/src/napakalaki/NumericBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author jose
 */
public class NumericBadConsequence extends BadConsequence {
    
    private int nVisibleTreasures , nHiddenTreasures;
    
    public NumericBadConsequence( String text , int levels , int nVisible , int nHidden){
        super(text,levels);
        nVisibleTreasures = nVisible;
        nHiddenTreasures = nHidden;
    }
    
    @Override
    public boolean isEmpty(){
        if ( nVisibleTreasures == 0 && nHiddenTreasures == 0 ){
            return true;
        }
        return false;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        if ( nVisibleTreasures > 0 ){
            nVisibleTreasures--;
        }
    }
    
    @Override
    public void substractHiddenTreasure( Treasure t ){
        if( nHiddenTreasures > 0 ){
            nHiddenTreasures--; 
        }
    }
    
    @Override
    public NumericBadConsequence adjustToFitTreasureLists( ArrayList<Treasure> v , ArrayList<Treasure> h ){
        int nv , nh;
        
        if( this.getnVisibleTreasures() == 0 ){
            nv = 0;
        }
        else if( this.getnVisibleTreasures() < v.size() ){
            nv = this.getnVisibleTreasures();
        }
        else{
            nv = v.size();
        }
        if ( this.getnHiddenTreasures() == 0 ){
            nh = 0;
        }
        else if( this.getnHiddenTreasures() < h.size() ){
            nh = this.getnHiddenTreasures();
        }
        else{
            nh = h.size();
        }
        NumericBadConsequence res = new NumericBadConsequence(this.getText(),0,nv,nh);
        return res;  
    }
    
    public int getnVisibleTreasures(){
        return nVisibleTreasures;
    }
    public int getnHiddenTreasures(){
        return nHiddenTreasures;
    }
    
    @Override
    public String toString(){
        return super.toString() + " NºTesoros visibles: " + Integer.toString(nVisibleTreasures) + " NºTesoros ocultos: " + Integer.toString(nHiddenTreasures);
    }
    
}
